### PR TITLE
Port to use Dune as build system

### PR DESCRIPTION
**Pull request by palmskog**
Using the regular coq_makefile for coq-dpdgraph is a lot of work for an essentially deprecated approach to building Coq plugins. Here is a quick port to Dune instead. The plugin and tools build fine, but two issues are not resolved yet:
1. The documentation needs to be updated to reflect that autoconf is no longer needed and give the key Dune commands.
2. The test suite needs to be ported to Dune.

The first issue is straightforward, but the second one is time-consuming without in-depth Dune knowledge that I currently lack. 

@ybertot perhaps we could merge this after the documentation is done and postpone the test suite port to some later point? We may be able to get some help from the Dune experts.

### CODE REVIEW

File: version.ml [hunk 1]
```diff
@@ -0,0 +1 @@
+let version = "1.0"
```
Comment (Alizter) [inline version.ml]: What is this for?

File: dune-project [hunk 2]
```diff
@@ -0,0 +1,4 @@
+(lang dune 2.7)
+(using coq 0.2)
+(name coq-dpdgraph)
+(cram enable)
```
Comment (Alizter) [inline dune-project]:
```suggestion
(cram enable)

```

File: dune [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ (name dpdusage)
+ (public_name dpdusage)
+ (libraries dpd_compute dpd_lex version)
+ (modules dpdusage)
+ (package coq-dpdgraph))
+
+(coq.theory
+ (name dpdgraph)
+ (package coq-dpdgraph)
+ (synopsis "Compute dependencies between Coq objects (definitions, theorems) and produce graphs")
```
Comment (Alizter) [inline dune]:
```suggestion
 (synopsis "Compute dependencies between Coq objects (definitions, theorems) and produce graphs.")
```

**Comment by Alizter** [comment]
Since the test-suite has been ported, what is left keeping this back?

File: dune (hunk 3 above)
Comment (palmskog) [reply dune]: According to opam, synopses are not supposed to end in a period, and in fact ending in a period raises a warning.

File: version.ml (hunk 1 above)
Comment (palmskog) [reply version.ml]: The CLI tools display a version number when listing help. I believe we have now decided that this is version 1.0 going forward.

**Comment by palmskog** [comment]
@Alizter the final refactorings that I have pending locally before I mark this to be ready are:
- have the test `.v` files explicitly rather than as defined inside cram `.t` files
- split into `src` and `theories` directory

**Comment by Alizter** [comment]
@palmskog I can split those .v files.

**Comment by palmskog** [comment]
I have already done the splitting locally, but I ran into the Dune issues discussed on Zulip. I'll take care of the final changes, and you can review when I mark as ready for review.

File: dune-project (hunk 2 above)
Comment (palmskog) [reply dune-project]: Not sure what the extra newline here would add, the convention in Coq-community is usually to skip them.

**Comment by palmskog** [comment]
@Alizter this PR is now ready for re-review. The only other changes I might do at this point (other those than suggested during review) are to documentation, which still talks about autoconf and similar.

File: .github/workflows/docker-action.yml [hunk 4]
```diff
@@ -17,7 +17,7 @@ jobs:
     strategy:
       matrix:
         image:
-          - 'coqorg/coq:dev'
+          - 'coqorg/coq:dev-ocaml-4.13.1-flambda'
```
Comment (Alizter) [inline .github/workflows/docker-action.yml]: Would it also be useful to test the currently release of Coq?

Comment (palmskog) [reply .github/workflows/docker-action.yml]: Not sure I follow. This is a plugin where the `coq-master` branch is rarely compatible with anything other than the last 3-4 commits to Coq `master`. If I add `coqorg/coq:8.15`, it's nearly guaranteed to fail. If we get a Docker image for the `v8.16` branch of Coq, the `coq-master` branch here is not likely to stay compatible with it for long (we should instead create the branch `coq-8.16` from the last commit known to work).

Comment (Alizter) [reply .github/workflows/docker-action.yml]: Yes, I forgot that the plugin has releases for the coq version anyhow.

**Comment by Alizter** [comment]
We still have the
```
File "_unknown_", line 1, characters 0-0:
*** Error: in file dpdgraph.v, could not find META.coq-dpdgraph.
```
issue. I am unsure if there is a work around other than doing `dune build` twice. I'll see what I can do.

**Comment by palmskog** [comment]
@Alizter I can't replicate that locally (I don't get the error anymore). Shouldn't that error show up in CI then? I don't see it here: https://github.com/coq-community/coq-dpdgraph/runs/6453619501?check_suite_focus=true#step:4:390

**Comment by Alizter** [comment]
@palmskog Seems my cache got corrupted from before, <s>I disabled cache now and it is working fine.</s>

Actually no. Let me see.

---

Seems to be an issue with coqdep. I suspect its because we are relying on the META file of the coq-dpdgraph package whilst generating rules that will be installed in the coq-dpdgraph package.

**Comment by Alizter** [comment]
OK I had a discussion with Rudi about this. I think barring Coq's poor design choice to have circular deps on META, the best solution here to avoid this problem is to have two separate packages. One for the plugin and one for the theory. We can later package them up further into coq-dpdgraph at the end for opam.

**Comment by palmskog** [comment]
@Alizter I don't think this is a workable solution. The "theory" here consists of exactly one line of code:
```coq
Declare ML Module "coq-dpdgraph.plugin".
```
coq-dpdgraph is currently a single package using autoconf + a regular Makefile that is part of the Coq Platform. To avoid the large overhead of maintaining two packages, I think we should then instead keep this PR open until the Dune issues with coqdep are sorted.

**Review by Alizter — APPROVED** [review]
Never mind, I hadn't checked out the PR properly and was stuck on an older version where the ml and v fiels where mixed. That was causing the coq.theory stanza to break. Now that everything is in the proper folder it looks like it is working as intended.

**Comment by palmskog** [comment]
@ybertot Ali and I believe this is ready to be merged now, but we would appreciate if you could take a look at the changes, in particular in the test suite. The key command you may want to run locally is:
```shell
dune runtest
```

One important thing to note is that if we merge this, we must update the build scripts in Coq's CI - the old build process scripted there will not work anymore.

**Comment by ybertot** [comment]
I have to run to a few things these days, I may not have time to look at it before the end of the week.  I will try to keep it on my agenda.

**Comment by palmskog** [comment]
@ybertot do you think we can rebase+merge this PR and switch to Dune for a release of dpdgraph for Coq 8.17 now that 8.17+rc1 is out? If you don't mind, I could do it.

**Comment by Alizter** [comment]
@palmskog Are we really ready to switch to Dune for this? I thought we were going to do less changes to the tests then we currently do. Also what is the rush for 8.17?

**Comment by palmskog** [comment]
There is no big rush, but if we can throw out the dependency on autoconf/autotools for good for the next Coq Platform release (in March 2023 I guess), that would be good for everyone involved. Since I believe we faithfully capture all existing tests, there is nothing in my view that stand in the way.

**Comment by ybertot** [comment]
Moving from autoconf to dune means moving from a system I don't understand fully to a system I barely understand.  Right now, I have administrative duties that make that I cannot guarantee reading the documentation of Dune before end of January, but it is good of you to have sent me a reminder.

If there is no hurry, I suggest we pospone my answer another 2 or 3 weeks, to see if I can gather enough information by then.

**Comment by palmskog** [comment]
Fine by me to postpone further PR discussion until February.

**Comment by Alizter** [comment]
@ybertot I have added more details to the README.md and a basic Makefile with things like:
```
make
make build (same as above)
make test (runs tests)
make promote (promotes differing output tests)
make install (installs)
```
The CI also now runs the test-suite.

**Comment by palmskog** [comment]
@ybertot @Alizter do we have any consensus here on a plan for merging the PR? The advantage with a Dune-based build is that many people in Coq-community could potentially do releases of coq-dpdgraph when required - there is no local preparation of a release archive needed, just a Git tag and package definition.

**Comment by Alizter** [comment]
@palmskog last I spoke to @ybertot he wanted to have a look at the readme in some more detail. So we are waiting for him.

**Comment by ybertot** [comment]
Thanks for the reminder.

**Comment by SkySkimmer** [comment]
>Using the regular coq_makefile for coq-dpdgraph is a lot of work for an essentially deprecated approach to building Coq plugins.

coq_makefile is not deprecated.

**Pull request merged**